src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.cpp

```diff
@@ -0,0 +1,237 @@
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map the Vulkan memory for read only!");
+        }
+
+        return (reinterpret_cast<uint8_t*>(pDestination) + readRangeOffset);
+    }
+
+    void VulkanGraphicsBuffer::Unmap()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        vkUnmapMemory(vulkanDevice, vulkanDeviceMemory);
+    }
+
+    void VulkanGraphicsBuffer::UnmapAndWrite()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        uint64_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+
+        size_t offset = GetOffset();
+        size_t size = (GetSize() + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult flushMappedMemoryRangesResult = vkFlushMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+        if (flushMappedMemoryRangesResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to flush the written changes to the Vulkan memory!");
+        }
+
+        vkUnmapMemory(vulkanDevice, vulkanDeviceMemory);
+    }
+
+    void VulkanGraphicsBuffer::UnmapAndWrite(size_t writtenRangeOffset, size_t writtenRangeLength)
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        uint64_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+
+        size_t offset = GetOffset() + writtenRangeOffset;
+        size_t size = (writtenRangeLength + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult flushMappedMemoryRangesResult = vkFlushMappedMemoryRanges(vulkanDevice, 1, &mappedMemoryRange);
+
+        if (flushMappedMemoryRangesResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to flush the written changes to the Vulkan memory!");
```

Comment:
Press F to pay respects to real exceptions.

Reply:
This will probably be the same as the other flush exception in VulkanGraphicsBuffer.